src/Microsoft.AspNet.OData.Shared/DeltaOfTStructuralType.cs

```diff
@@ -465,6 +519,51 @@ public void Patch(TStructuralType original)
             CopyChangedValues(original);
         }
 
+        /// <summary>
+        /// Overwrites the <paramref name="original"/> entity with the changes tracked by this Delta.
+        /// <remarks>The semantics of this operation are equivalent to a HTTP PATCH operation, hence the name.</remarks>
+        /// </summary>
+        /// <param name="original">The entity to be updated.</param>        
+        /// <param name="apiHandlerFactory">API Handler Factory</param>
+        internal void Patch(TStructuralType original, ODataAPIHandlerFactory apiHandlerFactory)
+        {
+            IODataAPIHandler apiHandler = apiHandlerFactory.GetHandler(_navigationPath);
+
+            Debug.Assert(apiHandler != null);
+
+            CopyChangedValues(original, apiHandler as ODataAPIHandler<TStructuralType>, apiHandlerFactory);            
+        }
+
+        /// <summary>
+        /// This is basically Patch on ODataId. This applies ODataId parsed Navigation paths, get the value identified by that and copy it on original object
+        /// </summary>    
+        private void ApplyODataId(TStructuralType original, ODataAPIHandlerFactory apiHandlerFactory)
+        {
+            IODataAPIHandler refapiHandler = apiHandlerFactory.GetHandler(ODataIdContainer.ODataIdNavigationPath);
+
+            if (refapiHandler != null)
+            {
+                ODataAPIHandler<TStructuralType> refapiHandlerOfT = refapiHandler as ODataAPIHandler<TStructuralType>;
+
+                Debug.Assert(refapiHandlerOfT != null);
+
+                TStructuralType referencedObj;
+                string error;
+
+                //Checking to get the referenced entity, get the properties and apply it on original object
+                if (refapiHandlerOfT.TryGet(ODataIdContainer.ODataIdNavigationPath.GetNavigationPathItems().Last().KeyProperties, out referencedObj, out error) == ODataAPIResponseStatus.Success)
```

Comment:
Is `ODataIdContainer.ODataIdNavigationPath.GetNavigationPathItems().Last()` guaranteed to return a non-null?